File: sudoku/sudoku_v1.dart
import 'dart:core';
import 'dart:io';

class Sudoku {
	var _R, _C;
	Sudoku() {
		_R = new List(324);
		_C = new List(729);
		for (int i = 0, r = 0; i < 9; ++i)
			for (int j = 0; j < 9; ++j)
				for (int k = 0; k < 9; ++k)
					_C[r++] = <int>[ 9 * i + j, (i~/3*3 + j~/3) * 9 + k + 81, 9 * i + k + 162, 9 * j + k + 243 ];
		for (int c = 0; c < 324; ++c) _R[c] = new List<int>();
		for (int r = 0; r < 729; ++r)
			for (int c2 = 0; c2 < 4; ++c2)
				_R[_C[r][c2]].add(r);
	}
	_update(sr, sc, r, v) {
		int min = 10, min_c = 0;
		for (int c2 = 0; c2 < 4; ++c2) sc[_C[r][c2]] += v<<7;
		for (int c2 = 0; c2 < 4; ++c2) {
			int r2, rr, cc2, c = _C[r][c2];
			if (v > 0) {
				for (r2 = 0; r2 < 9; ++r2) {
					if (sr[rr = _R[c][r2]]++ != 0) continue;
					for (cc2 = 0; cc2 < 4; ++cc2) {
						var cc = _C[rr][cc2];
						if (--sc[cc] < min) { min = sc[cc]; min_c = cc; }
					}
				}
			} else { // revert
				for (r2 = 0; r2 < 9; ++r2) {
					if (--sr[rr = _R[c][r2]] != 0) continue;
					var p = _C[rr];
					++sc[p[0]]; ++sc[p[1]]; ++sc[p[2]]; ++sc[p[3]];
				}
			}
		}
		return min<<16 | min_c; // return the col that has been modified and with the minimal available choices
	}
	solve(s_) {
		int j, r, c, r2, min, hints = 0;
		var sr = new List<int>(729), sc = new List<int>(324);
		var cr = new List<int>(81), cc = new List<int>(81), out = new List<int>(81), ret = [];
		for (r = 0; r < 729; ++r) sr[r] = 0;
		for (c = 0; c < 324; ++c) sc[c] = 9;
		for (int i = 0; i < 81; ++i) {
			int a = s_.charCodeAt(i) >= 49 && s_.charCodeAt(i) <= 57? s_.charCodeAt(i) - 49 : -1;
			if (a >= 0) _update(sr, sc, i * 9 + a, 1);
			if (a >= 0) ++hints;
			cr[i] = cc[i] = -1; out[i] = a + 1;
		}
		for (int i = 0, dir = 1, cand = 10<<16|0;;) {
			while (i >= 0 && i < 81 - hints) {
				if (dir == 1) {
					min = cand>>16; cc[i] = cand&0xffff;
					if (min > 1) {
						for (c = 0; c < 324; ++c) {
							if (sc[c] < min) {
								min = sc[c]; cc[i] = c;
								if (min <= 1) break;
							}
						}
					}
					if (min == 0 || min == 10) cr[i--] = dir = -1;
				}
				c = cc[i];
				if (dir == -1 && cr[i] >= 0) _update(sr, sc, _R[c][cr[i]], -1);
				for (r2 = cr[i] + 1; r2 < 9; ++r2)
					if (sr[_R[c][r2]] == 0) break;
				if (r2 < 9) {
					cand = _update(sr, sc, _R[c][r2], 1);
					cr[i++] = r2; dir = 1;
				} else cr[i--] = dir = -1;
			}
			if (i < 0) break;
			var y = new List<int>(81);
			for (j = 0; j < 81; ++j) y[j] = out[j];
			for (j = 0; j < i; ++j) { r = _R[cc[j]][cr[j]]; y[r~/9] = r%9 + 1; }
			ret.add(y);
			--i; dir = -1;
		}
		return ret;
	}
}

main()
{
	List<String> argv = new Options().arguments;
	var fp = new File(argv[0]);
	var lines = fp.readAsLinesSync(Encoding.ASCII);
	Sudoku s = new Sudoku();
	for (int i = 0; i < lines.length; ++i) {
		var ret = s.solve(lines[i]);
		print(ret[0]);
	};
}

File: sudoku/sudoku_v1.rs
#![feature(phase)]
#[phase(plugin, link)] extern crate log;

struct Sudoku {
	r: [[u16, ..9], ..324],
	c: [[u16, ..4], ..729]
}

impl Sudoku {
	pub fn new() -> Sudoku {
		let mut s = Sudoku { r: [[0, ..9], ..324], c: [[0, ..4], ..729] };
		let mut nr = [0, ..324];
		let mut r = 0;
		for i in range(0, 9) {
			for j in range(0, 9) {
				for k in range(0, 9) {
					s.c[r][0] = 9 * i + j;
					s.c[r][1] = (i/3*3 + j/3) * 9 + k + 81;
					s.c[r][2] = 9 * i + k + 162;
					s.c[r][3] = 9 * j + k + 243;
					r += 1;
				}
			}
		}
		for r in range(0, 729) {
			for c2 in range(0, 4) {
				let k = s.c[r][c2];
				s.r[k as uint][nr[k as uint]] = r as u16;
				nr[k as uint] += 1;
			}
		}
		return s;
	}
	fn forward(&self, sr: &mut [i8], sc: &mut [u8], c: uint, min: &mut u8, min_c: &mut u16) {
		for &rr in self.r[c].iter() {
			// Take a pointer to avoid repeated bounds checks
			let srrr = &mut sr[rr as uint];
			*srrr += 1;
			if *srrr == 1 {
				for &cc in self.c[rr as uint].iter() {
					// Take a pointer to avoid repeated bounds checks
					let sccc = &mut sc[cc as uint];
					*sccc -= 1;
					if *sccc < *min {
						*min = *sccc; *min_c = cc;
					}
				}
			}
		}
	}
	fn revert(&self, sr: &mut [i8], sc: &mut [u8], c: uint) {
		for &rr in self.r[c].iter() {
			// Take a pointer to avoid repeated bounds checks
			let srrr = &mut sr[rr as uint];
			*srrr -= 1;
			if *srrr == 0 {
				for &i in self.c[rr as uint].iter() {
					sc[i as uint] += 1;
				}
			}
		}
	}
	fn update(&self, sr: &mut [i8], sc: &mut [u8], r: uint, v: int) -> int {
		let mut min = 10;
		let mut min_c = 0;
		for &i in self.c[r].iter() {
			sc[i as uint] += (v<<7) as u8;
		}
		for &c in self.c[r].iter() {
			if v > 0 { // move forward
				self.forward(sr, sc, c as uint, &mut min, &mut min_c)
			} else {
				self.revert(sr, sc, c as uint)
			}
		}
		return (min as int)<<16 | min_c as int;
	}
	pub fn solve(&self, inp: &str) -> Vec<String> {
		let mut sc = [9u8, ..324];
		let mut sr = [0i8, ..729];
		let mut cr = [-1i8, ..81];
		let mut cc = [-1i16, ..81];
		let mut s = [0u, ..81];
		let mut s8 = [48u8, ..81];
		let mut hints = 0;
		for (i, c) in inp.chars().enumerate().take(81) {
			s[i] = -1;
			if c >= '1' && c <= '9' {
				s[i] = c as uint - '1' as uint;
				self.update(sr, sc, (i * 9 + s[i]), 1);
				hints += 1;
				s8[i] = c as u8;
			}
		}

		let mut ret: Vec<String> = Vec::new();
		let mut i = 0i;
		let mut dir = 1i;
		let mut cand: int = 10<<16|0;
		loop {
			while i >= 0 && i < 81 - hints {
				if dir == 1 {
					let mut min = (cand>>16) as u8;
					cc[i as uint] = (cand & 0xffff) as i16;
					if min > 1 {
						for (c, &v) in sc.iter().enumerate() {
							if v < min {
								min = v; cc[i as uint] = c as i16;
								if min <= 1 { break; }
							}
						}
					}
					if min == 0 || min == 10 {
						cr[i as uint] = -1; dir = -1; i -= 1;
						assert!(i >= 0);
					}
				}
				let c = cc[i as uint];
				if dir == -1 && cr[i as uint] >= 0 {
					self.update(sr, sc, self.r[c as uint][cr[i as uint] as uint] as uint, -1);
				}
				let mut tmp = 9i8;
				for r2 in range(cr[i as uint] + 1, 9) {
					if sr[self.r[c as uint][r2 as uint] as uint] == 0 {
						tmp = r2;
						break;
					}
				}
				if tmp < 9 {
					cand = self.update(sr, sc, self.r[c as uint][tmp as uint] as uint, 1);
					cr[i as uint] = tmp; dir = 1; i += 1;
				} else {
					cr[i as uint] = -1; dir = -1; i -= 1;
				}
			}
			if i < 0 { break; }
			for j in range(0, i as uint) {
				let r = self.r[cc[j] as uint][cr[j] as uint];
				s8[r as uint/9] = (r%9 + '1' as u16) as u8;
			}
			ret.push(String::from_utf8_lossy(s8).to_string());
			i -= 1; dir = -1;
		}
		return ret;
	}
}

fn main() {
	let s = Sudoku::new();
	for e_line in std::io::stdin().lines() {
		match e_line {
			Ok(line) => {
				let r = s.solve(line.as_slice());
				for l in r.iter() {
					std::io::println(l.as_slice());
				}
				std::io::println("");
			},
			Err(e) => {
				error!("error reading: {}", e);
				return;
			}
		}
	}
}
